File: blender_config_sync/backup_engine.py
"""
Blender 配置备份与恢复引擎
负责配置文件的打包、压缩、导入和导出
"""

import json
import shutil
import zipfile
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Optional, Any
from dataclasses import dataclass, asdict, field


@dataclass
class BackupManifest:
    """备份清单"""
    version: str = "1.0"
    created_at: str = ""
    source_blender_version: str = ""
    source_config_path: str = ""
    files: List[Dict] = field(default_factory=list)
    total_size: int = 0
    checksum: str = ""


@dataclass
class BackupResult:
    """备份操作结果"""
    success: bool
    backup_path: str = ""
    manifest: Optional[BackupManifest] = None
    message: str = ""
    warnings: List[str] = field(default_factory=list)


@dataclass
class RestoreResult:
    """恢复操作结果"""
    success: bool
    restored_files: int = 0
    skipped_files: int = 0
    errors: List[str] = field(default_factory=list)
    rollback_available: bool = False
    backup_path: str = ""


class BackupEngine:
    """配置备份引擎"""

    BACKUP_DIR_NAME = "backups"
    MANIFEST_FILENAME = "manifest.json"

    # 需要备份的关键文件/目录
    BACKUP_TARGETS = {
        'config': [
            'userpref.blend',
            'startup.blend',
            'bookmarks.txt',
            'recent-files.txt',
        ],
        'scripts': {
            'addons': 'addons',
            'addons_contrib': 'addons_contrib',
            'startup': 'startup',
            'presets/keyconfig': 'presets/keyconfig',
        }
    }

    def __init__(self, output_base_dir: Path = None):
        """
        初始化备份引擎

        Args:
            output_base_dir: 备份输出基础目录，默认为当前工作目录下的 backups/
        """
        if output_base_dir is None:
            output_base_dir = Path.cwd() / self.BACKUP_DIR_NAME

        self.output_base = output_base_dir
        self.output_base.mkdir(parents=True, exist_ok=True)

    def create_backup(self, config_path: Path, blender_version: str,
                      include_addons: bool = True, compression: int = zipfile.ZIP_DEFLATED,
                      backup_type: str = "") -> BackupResult:
        """
        创建配置备份

        Args:
            config_path: Blender 配置目录路径
            blender_version: Blender 版本号
            include_addons: 是否包含插件
            compression: 压缩类型
            backup_type: 备份类型标记（'source' 或 'target'），用于文件名区分

        Returns:
            BackupResult: 备份结果
        """
        result = BackupResult(success=False)

        if not config_path.exists():
            result.message = f"❌ 配置路径不存在: {config_path}"
            return result

        # 优先从上级目录读取版本号（Blender标准目录结构）
        # 例如 config_path = ~/.config/blender/4.2/config，则上级目录是 4.2
        parent_dir = config_path.parent
        version_from_parent = parent_dir.name if parent_dir.name else ""
        
        # 检查上级目录名是否是有效版本号格式（如 4.2, 3.6, 4.2.1）
        is_valid_version = False
        if version_from_parent:
            parts = version_from_parent.replace('.', '').split()
            is_valid_version = parts[0].isdigit() and len(parts) >= 1
        
        # 决定使用的版本号：优先上级目录，否则用传入的版本号
        if is_valid_version and version_from_parent.replace('.', '').isdigit():
            final_version = version_from_parent
        else:
            final_version = blender_version
        
        # 解析版本号，只保留主版本和次版本（如 4.2.1 -> 4.2）
        version_parts = final_version.split('.')
        version_short = f"{version_parts[0]}.{version_parts[1]}" if len(version_parts) >= 2 else final_version
        
        # 创建文件名: Blender_X.X_Portable_YYYYMMDD.zip
        timestamp = datetime.now().strftime('%Y%m%d')
        backup_filename = f'Blender_{version_short}_Portable_{timestamp}.zip'
        backup_path = self.output_base / backup_filename

        try:
            manifest = BackupManifest(
                created_at=datetime.now().isoformat(),
                source_blender_version=blender_version,
                source_config_path=str(config_path)
            )

            with zipfile.ZipFile(backup_path, 'w', compression=compression) as zf:
                config_dir = config_path / 'config'
                scripts_dir = config_path / 'scripts'

                # 备份 config 目录下的关键文件
                for filename in self.BACKUP_TARGETS['config']:
                    file_path = config_dir / filename
                    if file_path.exists():
                        file_info = self._add_file_to_zip(zf, file_path, f'config/{filename}')
                        manifest.files.append(file_info)
                    else:
                        result.warnings.append(f"⚠️ 文件不存在: {filename}")

                # 备份 scripts 目录（如果需要）
                if include_addons and scripts_dir.exists():
                    for dir_name, relative_name in self.BACKUP_TARGETS['scripts'].items():
                        source_dir = scripts_dir / dir_name
                        if source_dir.exists():
                            added_files = self._add_directory_to_zip(
                                zf, source_dir, f'scripts/{relative_name}'
                            )
                            manifest.files.extend(added_files)

                # 计算总大小
                manifest.total_size = sum(f['size'] for f in manifest.files)

            # 写入清单文件
            manifest_data = json.dumps(asdict(manifest), indent=2, ensure_ascii=False)
            self._update_manifest_in_zip(backup_path, manifest_data)
            manifest.checksum = self._calculate_file_hash(backup_path)

            result.success = True
            result.backup_path = str(backup_path)
            result.manifest = manifest
            result.message = f"✅ 备份成功！共 {len(manifest.files)} 个文件，总计 {manifest.total_size / 1024:.1f} KB"

        except Exception as e:
            result.message = f"❌ 备份失败: {str(e)}"
            if backup_path.exists():
                backup_path.unlink()

        return result

    def _add_file_to_zip(self, zf: zipfile.ZipFile, file_path: Path, arcname: str) -> Dict:
        """添加单个文件到 ZIP"""
        stat = file_path.stat()
        zf.write(file_path, arcname)

        return {
            'path': arcname,
            'size': stat.st_size,
            'modified': datetime.fromtimestamp(stat.st_mtime).isoformat(),
            'type': 'file'
        }

    def _add_directory_to_zip(self, zf: zipfile.ZipFile, dir_path: Path, base_arcname: str) -> List[Dict]:
        """添加目录及其内容到 ZIP"""
        added_files = []

        for item in dir_path.rglob('*'):
            if item.is_file():
                rel_path = item.relative_to(dir_path)
                arcname = f'{base_arcname}/{rel_path}'
                added_files.append(self._add_file_to_zip(zf, item, arcname))

        return added_files

    def _update_manifest_in_zip(self, zip_path: Path, manifest_json: str):
        """更新 ZIP 文件中的清单"""
        temp_path = zip_path.with_suffix('.tmp.zip')

        with zipfile.ZipFile(zip_path, 'r') as zin:
            with zipfile.ZipFile(temp_path, 'w', zipfile.ZIP_DEFLATED) as zout:
                zout.writestr(self.MANIFEST_FILENAME, manifest_json)
                for item in zin.infolist():
                    if item.filename != self.MANIFEST_FILENAME:
                        zout.writestr(item, zin.read(item.filename))

        temp_path.replace(zip_path)

    def restore_backup(self, backup_path: Path, target_config_path: Path,
                       overwrite: bool = False, create_backup_first: bool = True) -> RestoreResult:
        """
        从备份恢复配置

        Args:
            backup_path: 备份文件路径
            target_config_path: 目标配置目录
            overwrite: 是否覆盖已存在的文件
            create_backup_first: 是否在恢复前先备份目标

        Returns:
            RestoreResult: 恢复结果
        """
        result = RestoreResult(success=False)

        if not backup_path.exists():
            result.errors.append(f"❌ 备份文件不存在: {backup_path}")
            return result

        if not target_config_path.exists():
            target_config_path.mkdir(parents=True, exist_ok=True)

        try:
            # 可选：先备份现有配置
            if create_backup_first:
                pre_restore_backup = self._pre_restore_backup(target_config_path)
                if pre_restore_backup:
                    result.rollback_available = True
                    result.backup_path = pre_restore_backup

            # 读取清单
            manifest = self.read_manifest(backup_path)
            if not manifest:
                result.errors.append("❌ 无法读取备份清单")
                return result

            restored_count = 0
            skipped_count = 0

            with zipfile.ZipFile(backup_path, 'r') as zf:
                for file_info in zf.infolist():
                    if file_info.filename == self.MANIFEST_FILENAME:
                        continue

                    target_file = target_config_path / file_info.filename

                    # 检查是否覆盖
                    if target_file.exists() and not overwrite:
                        skipped_count += 1
                        continue

                    # 确保父目录存在
                    target_file.parent.mkdir(parents=True, exist_ok=True)

                    # 提取文件
                    zf.extract(file_info, target_config_path)
                    restored_count += 1

            result.success = True
            result.restored_files = restored_count
            result.skipped_files = skipped_count

        except Exception as e:
            result.errors.append(f"❌ 恢复失败: {str(e)}")

        return result

    def _pre_restore_backup(self, config_path: Path) -> Optional[str]:
        """恢复前自动备份"""
        try:
            timestamp = datetime.now().strftime('%Y%m%d_%H%M%S')
            pre_backup_filename = f'pre_restore_{timestamp}.zip'

            # 尝试检测版本号
            version = config_path.name if config_path.name else 'unknown'

            result = self.create_backup(
                config_path=config_path,
                blender_version=version,
                include_addons=True
            )

            if result.success:
                # 重命名备份文件
                new_path = self.output_base / pre_backup_filename
                Path(result.backup_path).rename(new_path)
                return str(new_path)

        except Exception:
            pass

        return None

    def read_manifest(self, backup_path: Path) -> Optional[BackupManifest]:
        """读取备份文件的清单"""
        try:
            with zipfile.ZipFile(backup_path, 'r') as zf:
                if self.MANIFEST_FILENAME not in zf.namelist():
                    return None

                manifest_data = zf.read(self.MANIFEST_FILENAME).decode('utf-8')
                data = json.loads(manifest_data)

                return BackupManifest(**data)
        except Exception:
            return None

    def list_backups(self) -> List[Dict]:
        """
        列出所有可用的备份

        Returns:
            List[Dict]: 备份信息列表
        """
        backups = []

        if not self.output_base.exists():
            return backups

        for backup_file in sorted(self.output_base.glob('*.zip'), reverse=True):
            manifest = self.read_manifest(backup_file)
            size_mb = backup_file.stat().st_size / (1024 * 1024)

            backup_info = {
                'filename': backup_file.name,
                'path': str(backup_file),
                'size_mb': round(size_mb, 2),
                'created_at': manifest.created_at if manifest else '未知',
                'blender_version': manifest.source_blender_version if manifest else '未知',
                'files_count': len(manifest.files) if manifest else 0,
            }

            backups.append(backup_info)

        return backups

    def delete_backup(self, backup_path: Path) -> bool:
        """删除指定的备份"""
        try:
            if backup_path.exists() and backup_path.suffix == '.zip':
                backup_path.unlink()
                return True
        except Exception:
            pass
        return False

    @staticmethod
    def _calculate_file_hash(file_path: Path) -> str:
        """计算文件哈希值"""
        import hashlib
        sha256 = hashlib.sha256()
        with open(file_path, 'rb') as f:
            for chunk in iter(lambda: f.read(8192), b''):
                sha256.update(chunk)
        return sha256.hexdigest()


def main():
    """测试用例"""
    from blender_config_sync.path_manager import BlenderPathManager

    print("=" * 70)
    print("🎨 Blender Config Sync - 备份引擎测试")
    print("=" * 70)

    manager = BlenderPathManager()
    versions = manager.detect_installed_versions()

    if not versions:
        print("\n❌ 未找到 Blender 安装")
        return

    latest = versions[0]
    print(f"\n📌 使用版本: Blender {latest.version}")
    print(f"📁 配置路径: {latest.config_path}\n")

    engine = BackupEngine()

    # 创建备份
    print("📦 正在创建备份...")
    result = engine.create_backup(
        config_path=latest.config_path,
        blender_version=latest.version,
        include_addons=True
    )

    print(f"\n{result.message}")

    if result.warnings:
        print("\n⚠️ 警告:")
        for warning in result.warnings:
            print(f"   • {warning}")

    if result.success:
        print(f"\n💾 备份文件位置: {result.backup_path}")
        print(f"📊 清单信息:")
        print(f"   • Blender 版本: {result.manifest.source_blender_version}")
        print(f"   • 创建时间: {result.manifest.created_at}")
        print(f"   • 文件数量: {len(result.manifest.files)}")
        print(f"   • 总大小: {result.manifest.total_size / 1024:.1f} KB")
        print(f"   • 校验和: {result.manifest.checksum[:16]}...")

        # 列出所有备份
        print("\n📋 所有备份记录:")
        backups = engine.list_backups()
        for i, backup in enumerate(backups[:5], 1):
            print(f"   {i}. {backup['filename']:<45} "
                  f"{backup['size_mb']:>6.2f} MB  "
                  f"{backup['created_at'][:16]}  "
                  f"Blender {backup['blender_version']}")


if __name__ == '__main__':
    main()

File: src-tauri/src/backup_engine.rs
// backup_engine.rs - Blender 配置备份与恢复引擎

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::fs;
use std::io::{Read, Write, Seek};
use std::path::PathBuf;
use walkdir::WalkDir;
use zip::{ZipArchive, ZipWriter, write::FileOptions, CompressionMethod};

#[derive(Debug, Serialize, Deserialize)]
pub struct BackupManifest {
    pub version: String,
    pub created_at: String,
    pub source_blender_version: String,
    pub source_config_path: String,
    pub files: Vec<FileInfo>,
    pub total_size: u64,
    pub checksum: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct FileInfo {
    pub path: String,
    pub size: u64,
    pub modified: String,
    pub type_: String, // "file" or "dir"
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BackupResult {
    pub success: bool,
    pub backup_path: Option<String>,
    pub manifest: Option<BackupManifest>,
    pub message: String,
    pub warnings: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct RestoreResult {
    pub success: bool,
    pub restored_files: usize,
    pub skipped_files: usize,
    pub errors: Vec<String>,
    pub rollback_available: bool,
    pub backup_path: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct BackupInfo {
    pub filename: String,
    pub path: String,
    pub size_mb: f64,
    pub created_at: String,
    pub blender_version: String,
    pub source_config_path: String,
    pub files_count: usize,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SyncResult {
    pub success_count: usize,
    pub failed_count: usize,
    pub skipped_count: usize,
    pub errors: Vec<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SyncItemInput {
    pub category: String,
    pub item_type: String,
    pub name: String,
    pub action: String,
}

const MANIFEST_FILENAME: &str = "manifest.json";
const MANIFEST_VERSION: &str = "1.0";

/// 获取备份目录
fn get_backup_dir() -> PathBuf {
    dirs::config_dir()
        .unwrap_or_else(|| PathBuf::from("."))
        .join("blender-config-sync")
        .join("backups")
}

/// 创建配置备份
pub fn create_backup(config_path: &str, version: &str, include_addons: bool, output_dir: Option<&str>) -> std::io::Result<BackupResult> {
    let config_path_buf = PathBuf::from(config_path);
    if !config_path_buf.exists() {
        return Ok(BackupResult {
            success: false,
            backup_path: None,
            manifest: None,
            message: format!("配置路径不存在: {}", config_path),
            warnings: Vec::new(),
        });
    }

    // 优先使用用户指定的输出目录，否则使用默认目录
    let backup_dir = match output_dir {
        Some(dir) => PathBuf::from(dir),
        None => get_backup_dir(),
    };
    fs::create_dir_all(&backup_dir)?;

    // 解析版本号，只保留主版本和次版本
    let version_short = version.split('.')
        .take(2)
        .collect::<Vec<_>>()
        .join(".");

    let timestamp: DateTime<Utc> = Utc::now();
    let timestamp_str = timestamp.format("%Y%m%d").to_string();
    let backup_filename = format!("Blender_{}_Portable_{}.zip", version_short, timestamp_str);
    let backup_path = backup_dir.join(&backup_filename);

    let mut warnings = Vec::new();
    let mut manifest = BackupManifest {
        version: MANIFEST_VERSION.to_string(),
        created_at: timestamp.to_rfc3339(),
        source_blender_version: version.to_string(),
        source_config_path: config_path.to_string(),
        files: Vec::new(),
        total_size: 0,
        checksum: String::new(),
    };

    // 创建 ZIP 文件
    let file = fs::File::create(&backup_path)?;
    let mut zip_writer = ZipWriter::new(file);
    let options = FileOptions::<()>::default()
        .compression_method(CompressionMethod::Deflated);

    let config_dir = config_path_buf.join("config");
    let scripts_dir = config_path_buf.join("scripts");

    // 备份 config 目录下的关键文件
    let config_files = ["userpref.blend", "startup.blend", "bookmarks.txt", "recent-files.txt"];
    for filename in config_files.iter() {
        let file_path = config_dir.join(filename);
        if file_path.exists() {
            let info = add_file_to_zip(&mut zip_writer, &file_path, format!("config/{}", filename), options)?;
            manifest.files.push(info);
        } else {
            warnings.push(format!("文件不存在: {}", filename));
        }
    }

    // 备份 scripts 目录（如果需要）
    if include_addons && scripts_dir.exists() {
        let script_dirs = [
            ("addons", "addons"),
            ("addons_contrib", "addons_contrib"),
            ("startup", "startup"),
            ("presets/keyconfig", "presets/keyconfig"),
        ];

        for (src_name, arc_name) in script_dirs.iter() {
            let source_dir = scripts_dir.join(src_name);
            if source_dir.exists() {
                let added_files = add_directory_to_zip(&mut zip_writer, &source_dir, format!("scripts/{}", arc_name), options)?;
                manifest.files.extend(added_files);
            }
        }
    }

    // 计算总大小
    manifest.total_size = manifest.files.iter().map(|f| f.size).sum();

    // 写入 manifest
    let manifest_json = serde_json::to_string_pretty(&manifest)?;
    zip_writer.start_file(MANIFEST_FILENAME, options)?;
    zip_writer.write_all(manifest_json.as_bytes())?;

    zip_writer.finish()?;

    // 计算备份文件的哈希值
    manifest.checksum = calculate_file_hash(&backup_path)?;

    // 在移动 manifest 之前获取需要的信息
    let files_count = manifest.files.len();
    let total_size_kb = manifest.total_size / 1024;

    Ok(BackupResult {
        success: true,
        backup_path: Some(backup_path.to_string_lossy().to_string()),
        manifest: Some(manifest),
        message: format!("备份成功！共 {} 个文件，总计 {} KB", files_count, total_size_kb),
        warnings,
    })
}

/// 添加单个文件到 ZIP
fn add_file_to_zip<W: Write + Seek>(
    zip_writer: &mut ZipWriter<W>,
    file_path: &PathBuf,
    arc_name: String,
    options: FileOptions<()>,
) -> std::io::Result<FileInfo> {
    let metadata = fs::metadata(file_path)?;
    let modified: DateTime<Utc> = metadata.modified()?.into();

    zip_writer.start_file(&arc_name, options)?;
    let mut file = fs::File::open(file_path)?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;
    zip_writer.write_all(&buffer)?;

    Ok(FileInfo {
        path: arc_name,
        size: metadata.len(),
        modified: modified.to_rfc3339(),
        type_: "file".to_string(),
    })
}

/// 添加目录及其内容到 ZIP
fn add_directory_to_zip<W: Write + Seek>(
    zip_writer: &mut ZipWriter<W>,
    dir_path: &PathBuf,
    base_arc_name: String,
    options: FileOptions<()>,
) -> std::io::Result<Vec<FileInfo>> {
    let mut added_files = Vec::new();

    for entry in WalkDir::new(dir_path).into_iter().filter_map(|e| e.ok()) {
        if entry.file_type().is_file() {
            let rel_path = entry.path().strip_prefix(dir_path)
                .map_err(|e| std::io::Error::other(e.to_string()))?;
            let arc_name = format!("{}/{}", base_arc_name, rel_path.to_string_lossy());
            let info = add_file_to_zip(zip_writer, &entry.path().to_path_buf(), arc_name, options)?;
            added_files.push(info);
        }
    }

    Ok(added_files)
}

/// 计算文件哈希值
fn calculate_file_hash(path: &PathBuf) -> std::io::Result<String> {
    use sha2::{Sha256, Digest};
    let mut file = fs::File::open(path)?;
    let mut hasher = Sha256::new();
    let mut buffer = [0u8; 8192];

    while file.read(&mut buffer[..])? > 0 {
        hasher.update(&buffer);
    }

    Ok(hex::encode(hasher.finalize()))
}

/// 诊断信息：返回当前扫描的目录和找到的 ZIP 文件
pub fn diagnose_backups(backup_dir_override: Option<&str>) -> std::io::Result<serde_json::Value> {
    let backup_dir = match backup_dir_override {
        Some(dir) => PathBuf::from(dir),
        None => get_backup_dir(),
    };
    let mut result = serde_json::Map::new();
    result.insert("scanned_dir".to_string(), serde_json::Value::String(backup_dir.to_string_lossy().to_string()));
    result.insert("dir_exists".to_string(), serde_json::Value::Bool(backup_dir.exists()));

    let mut zip_files = Vec::new();
    if backup_dir.exists() {
        for entry in fs::read_dir(&backup_dir)? {
            let entry = entry?;
            let path = entry.path();
            let ext = path.extension().map(|e| e.to_string_lossy().to_string()).unwrap_or_default();
            if ext == "zip" {
                let metadata = fs::metadata(&path)?;
                let mut file_info = serde_json::Map::new();
                file_info.insert("name".to_string(), serde_json::Value::String(path.file_name().unwrap_or_default().to_string_lossy().to_string()));
                file_info.insert("size_bytes".to_string(), serde_json::Value::Number(serde_json::Number::from(metadata.len())));
                zip_files.push(serde_json::Value::Object(file_info));
            }
        }
    }
    result.insert("zip_files".to_string(), serde_json::Value::Array(zip_files.clone()));
    result.insert("zip_count".to_string(), serde_json::Value::Number(serde_json::Number::from(zip_files.len())));

    // 测试读取第一个 ZIP 的 manifest
    if let Some(first) = zip_files.first() {
        let name = first.get("name").and_then(|v| v.as_str()).unwrap_or("");
        if !name.is_empty() {
            let zip_path = backup_dir.join(name);
            let manifest_ok = read_manifest(&zip_path).is_ok();
            result.insert("first_manifest_ok".to_string(), serde_json::Value::Bool(manifest_ok));
        }
    }

    Ok(serde_json::Value::Object(result))
}

/// 列出所有备份
pub fn list_backups(backup_dir_override: Option<&str>) -> std::io::Result<Vec<BackupInfo>> {
    let backup_dir = match backup_dir_override {
        Some(dir) => PathBuf::from(dir),
        None => get_backup_dir(),
    };
    if !backup_dir.exists() {
        return Ok(Vec::new());
    }

    let mut backups = Vec::new();

    for entry in fs::read_dir(&backup_dir)? {
        let entry = entry?;
        let path = entry.path();

        if path.extension().map(|e| e == "zip").unwrap_or(false) {
            let metadata = fs::metadata(&path)?;
            let size_mb = metadata.len() as f64 / (1024.0 * 1024.0);

            let manifest = read_manifest(&path).ok();
            let filename = path.file_name().unwrap().to_string_lossy().to_string();

            backups.push(BackupInfo {
                filename,
                path: path.to_string_lossy().to_string(),
                size_mb: (size_mb * 100.0).round() / 100.0,
                created_at: manifest.as_ref().map(|m| m.created_at.clone()).unwrap_or_else(|| "未知".to_string()),
                blender_version: manifest.as_ref().map(|m| m.source_blender_version.clone()).unwrap_or_else(|| "未知".to_string()),
                source_config_path: manifest.as_ref().map(|m| m.source_config_path.clone()).unwrap_or_default(),
                files_count: manifest.as_ref().map(|m| m.files.len()).unwrap_or(0),
            });
        }
    }

    // 按创建时间排序（从新到旧）
    backups.sort_by(|a, b| b.created_at.cmp(&a.created_at));

    Ok(backups)
}

/// 读取备份文件中的 manifest
fn read_manifest(backup_path: &PathBuf) -> std::io::Result<BackupManifest> {
    let file = fs::File::open(backup_path)?;
    let mut archive = ZipArchive::new(file)?;

    let mut manifest_file = archive.by_name(MANIFEST_FILENAME)?;
    let mut content = String::new();
    manifest_file.read_to_string(&mut content)?;

    let manifest: BackupManifest = serde_json::from_str(&content)?;
    Ok(manifest)
}

/// 删除备份
pub fn delete_backup(path: &str) -> std::io::Result<bool> {
    let path_buf = PathBuf::from(path);
    if path_buf.exists() && path_buf.extension().map(|e| e == "zip").unwrap_or(false) {
        fs::remove_file(&path_buf)?;
        Ok(true)
    } else {
        Ok(false)
    }
}

/// 从备份恢复配置
pub fn restore_backup(backup_path: &str, target_path: &str, overwrite: bool) -> std::io::Result<RestoreResult> {
    let backup_path_buf = PathBuf::from(backup_path);
    let target_path_buf = PathBuf::from(target_path);

    if !backup_path_buf.exists() {
        return Ok(RestoreResult {
            success: false,
            restored_files: 0,
            skipped_files: 0,
            errors: vec![format!("备份文件不存在: {}", backup_path)],
            rollback_available: false,
            backup_path: None,
        });
    }

    if !target_path_buf.exists() {
        fs::create_dir_all(&target_path_buf)?;
    }

    let file = fs::File::open(&backup_path_buf)?;
    let mut archive = ZipArchive::new(file)?;

    let mut restored_count = 0usize;
    let mut skipped_count = 0usize;
    let errors: Vec<String> = Vec::new();

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)?;
        let filename = file.name().to_string();

        if filename == MANIFEST_FILENAME {
            continue;
        }

        let target_file = target_path_buf.join(&filename);

        if target_file.exists() && !overwrite {
            skipped_count += 1;
            continue;
        }

        // 确保父目录存在
        if let Some(parent) = target_file.parent() {
            fs::create_dir_all(parent)?;
        }

        // 提取文件
        let mut output = fs::File::create(&target_file)?;
        std::io::copy(&mut file, &mut output)?;
        restored_count += 1;
    }

    Ok(RestoreResult {
        success: true,
        restored_files: restored_count,
        skipped_files: skipped_count,
        errors,
        rollback_available: false,
        backup_path: None,
    })
}

/// 同步选中项到目标
pub fn sync_items(items: &[SyncItemInput], source_path: &str, target_path: &str) -> std::io::Result<SyncResult> {
    let source_path_buf = PathBuf::from(source_path);
    let target_path_buf = PathBuf::from(target_path);

    let mut success_count = 0usize;
    let mut failed_count = 0usize;
    let mut skipped_count = 0usize;
    let mut errors = Vec::new();

    for item in items {
        let result = sync_single_item(item, &source_path_buf, &target_path_buf);

        match result {
            Ok(true) => success_count += 1,
            Ok(false) => skipped_count += 1,
            Err(e) => {
                failed_count += 1;
                errors.push(format!("{}: {}", item.name, e));
            }
        }
    }

    Ok(SyncResult {
        success_count,
        failed_count,
        skipped_count,
        errors,
    })
}

/// 同步单个项目
fn sync_single_item(item: &SyncItemInput, source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    match item.category.as_str() {
        "bookmarks" => sync_bookmarks(source, target),
        "addons" => sync_addon(&item.name, source, target),
        "preferences" => sync_preferences(source, target),
        "presets" => sync_preset(&item.name, &item.item_type, source, target),
        "startup_scripts" => sync_startup_script(&item.name, source, target),
        "folder_file" => sync_generic_file_or_dir(&item.name, source, target),
        _ => Ok(false),
    }
}

fn sync_bookmarks(source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src_file = source.join("config").join("bookmarks.txt");
    let dst_file = target.join("config").join("bookmarks.txt");

    if src_file.exists() {
        fs::create_dir_all(dst_file.parent().unwrap())?;
        fs::copy(&src_file, &dst_file)?;
        Ok(true)
    } else {
        Ok(false)
    }
}

fn sync_addon(name: &str, source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src_addon = source.join("scripts").join("addons").join(name);
    let dst_addon = target.join("scripts").join("addons").join(name);

    if src_addon.exists() {
        fs::create_dir_all(dst_addon.parent().unwrap())?;

        if src_addon.is_dir() {
            if dst_addon.exists() {
                fs::remove_dir_all(&dst_addon)?;
            }
            copy_dir_all(&src_addon, &dst_addon)?;
        } else {
            fs::copy(&src_addon, &dst_addon)?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

/// 复制整个目录
fn copy_dir_all(src: &PathBuf, dst: &PathBuf) -> std::io::Result<()> {
    fs::create_dir_all(dst)?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let ty = entry.file_type()?;
        let src_path = entry.path();
        let dst_path = dst.join(entry.file_name());

        if ty.is_dir() {
            copy_dir_all(&src_path, &dst_path)?;
        } else {
            fs::copy(&src_path, &dst_path)?;
        }
    }
    Ok(())
}

fn sync_preferences(source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src_file = source.join("config").join("userpref.blend");
    let dst_file = target.join("config").join("userpref.blend");

    if src_file.exists() {
        fs::create_dir_all(dst_file.parent().unwrap())?;
        fs::copy(&src_file, &dst_file)?;
        Ok(true)
    } else {
        Ok(false)
    }
}

fn sync_preset(name: &str, preset_type: &str, source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src_preset = source.join("scripts").join("presets").join(preset_type).join(name);
    let dst_preset = target.join("scripts").join("presets").join(preset_type).join(name);

    if src_preset.exists() {
        fs::create_dir_all(dst_preset.parent().unwrap())?;

        if src_preset.is_dir() {
            if dst_preset.exists() {
                fs::remove_dir_all(&dst_preset)?;
            }
            copy_dir_all(&src_preset, &dst_preset)?;
        } else {
            fs::copy(&src_preset, &dst_preset)?;
        }
        Ok(true)
    } else {
        Ok(false)
    }
}

fn sync_startup_script(name: &str, source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src_file = source.join("scripts").join("startup").join(name);
    let dst_file = target.join("scripts").join("startup").join(name);

    if src_file.exists() {
        fs::create_dir_all(dst_file.parent().unwrap())?;
        fs::copy(&src_file, &dst_file)?;
        Ok(true)
    } else {
        Ok(false)
    }
}

/// 通用文件/目录同步（用于文件夹差异模块）
fn sync_generic_file_or_dir(rel_path: &str, source: &PathBuf, target: &PathBuf) -> std::io::Result<bool> {
    let src = source.join(rel_path);
    let dst = target.join(rel_path);

    if !src.exists() {
        return Ok(false);
    }

    if src.is_dir() {
        if dst.exists() {
            fs::remove_dir_all(&dst)?;
        }
        copy_dir_all(&src, &dst)?;
    } else {
        if let Some(parent) = dst.parent() {
            fs::create_dir_all(parent)?;
        }
        fs::copy(&src, &dst)?;
    }

    Ok(true)
}